### PR TITLE
Document Lock+Close and trivial Lock test.

### PR DESCRIPTION
**Pull request by taruti**

### CODE REVIEW

**Review by strib — APPROVED** [review]

File: fs.go [hunk 1]
```diff
@@ -136,9 +136,12 @@ type File interface {
 	io.Seeker
 	io.Closer
 	// Lock locks the file like e.g. flock. It protects against access from
-	// other processes.
+	// other processes. The lock will be released by a call to Unlock or closing
+	// the file.
+	// Recursive locking (trying to lock the same file multiple times) or having
+	// multiple handles to a file may result in implementation specific behaviour.
```
Comment (songgao) [inline fs.go:142]: Our implementation expires the lock after a certain time, so this is a little bit different. But I'm not sure how to describe that in context of file locking/unlocking.

**Comment by songgao** [comment]
Sorry for bringing it up so late, but thinking about it now, I think Lock/Unlock isn't the right semantics for what we actually do with git. What do you all think if we have following instead? It's basically the same thing but just with slightly modified semantics.

```
	// StartMutuallyExclusive marks the start of a mutually exclusive
	// operation. In detail:
	// 1) It flushes blocks of any writes that have happened,
	//    but do not actually commits them. In other words, the changes are not
	//    effective, but it will be much faster to make them effective in a
	//    future time.
	// 2) In case of a network file system, it syncs with remote server.
	// 3) It makes sure a lock on the file is held. That is, it makes sure
	//    attempts to lock the file from any other process will fail. However,
	//    the lock may expire after a while, when other processes can grab the
	//    lock again.
	StartMutuallyExclusive() error
	// EndMutuallyExclusive marks the end of a mututally exclusive operation.
	// In detail, it flushes all writes, and make changes effective, but only
	// if the lock has been grabbed previously by StartMutuallyExclusive(), and
	// has not expired. If StartMutuallyExclusive() wasn't called before, or
	// the lock has expired, EndMutuallyExclusive returns an error. If all
	// writes are flushed successfully, the lock is automatically released.
	EndMutuallyExclusive() error
```

**Comment by strib** [comment]
@songgao: I think Lock/Unlock is the right interface from a billy perspective, and maps more correctly to actual FS implementation.  "Flushing blocks" has nothing to do with most implementations of a billy.File, or with how go-git will use it.  We can put your documentation in our own implementation, but I don't think it makes sense in this library.

**Comment by songgao** [comment]
In addition to lock expiration issue, in our implementation, `Unlock` would fail if at the time of flushing the lock has expired. But I'm not sure it makes sense for go-git to fail if `Unlock` returns an error? If `Unlock` actually checks the lock and flushes the write, it just doesn't seem to be the right semantics.

What if we name them `StartMutuallyExclusive()` and `EndMutuallyExclusive()`, but do not mention the flushing details?

**Comment by strib** [comment]
I disagree, I think go-git should fail whenever `Unlock()` returns an error.  It should never happen on a local FS, so it would be an extraordinary event.  Same thing with an error on `Close()` (which is where the error would appear to go-git in practice, it seems).

I think naming these anything other than Lock/Unlock greatly reduces our chances of getting accepted into go-billy and go-git.  Unless it would actually change the way go-git uses them (see @taruti's PR), but I don't think what you propose would change the go-git implementation at all.

**Comment by songgao** [comment]
OK I'm convinced; sounds good! Let's stick to Lock/Unlock then.

**Pull request merged**